File: 2048 GAME PROJECT/c++ programming/quicksort.cpp
#include<iostream>
using namespace std;

void swap(int *a,int *b){
	int c;
	c=*a;
	*a=*b;
	*b=c;
}



int partition(int a[],int low,int high){
	int pivot=a[high];
	int i=low-1;
	for(int j=low;j<=high-1;j++){
		if(a[j]<pivot)
		{
		
		i++;
		swap(&a[i],&a[j]);
	}
	}
	swap(&a[i+1],&a[high]);
	
	return i+1;
}

void quicksort(int a[],int low,int high){
	if(low<high){
		int pi=partition(a,low,high);
	
		quicksort(a,low,pi-1);
		quicksort(a,pi+1,high);
	}
}

int main(){
	int n;
	int i;
	cout<<"enter number of elements in array:"<<endl;
	cin>>n;
	int a[n];
	for(i=0;i<n;i++){
		cin>>a[i];
		
	}
	quicksort(a,0,n-1);
	cout<<"sorted array:"<<endl;
	for(int i=0;i<n;i++){
		cout<<a[i];
	}
	return 0;
}

File: Java_Programs_for_beginners/Quick sort.java
// Quick sort in Java

import java.util.Arrays;

class QuickSort {

    // Function to partition the array on the basis of pivot element
    int partition(int array[], int low, int high) {

        // Select the pivot element
        int pivot = array[high];
        int i = (low - 1);

        // Put the elements smaller than pivot on the left and
        // greater than pivot on the right of pivot
        for (int j = low; j < high; j++) {
            if (array[j] <= pivot) {
                i++;
                int temp = array[i];
                array[i] = array[j];
                array[j] = temp;
            }
        }
        int temp = array[i + 1];
        array[i + 1] = array[high];
        array[high] = temp;
        return (i + 1);
    }

    void quickSort(int array[], int low, int high) {
        if (low < high) {

            // Select pivot position and put all the elements smaller
            // than pivot on left and greater than pivot on right
            int pi = partition(array, low, high);

            // Sort the elements on the left of pivot
            quickSort(array, low, pi - 1);

            // Sort the elements on the right of pivot
            quickSort(array, pi + 1, high);
        }
    }

    // Driver code
    public static void main(String args[]) {
        int[] data = {8, 7, 2, 1, 0, 9, 6};
        int size = data.length;
        QuickSort qs = new QuickSort();
        qs.quickSort(data, 0, size - 1);
        System.out.println("Sorted Array in Ascending Order: ");
        System.out.println(Arrays.toString(data));
    }
}
